ament_copyright/ament_copyright/__init__.py

```diff
@@ -12,31 +12,354 @@
+            print("Could not find copyright year in file '%s' in line [%s]" %
+                  (filename, line), file=sys.stderr)
+            continue
+        copyright_year = match.group(1)
+
+        # skip if copyright year is already up-to-date
+        if current_year in copyright_year:
+            continue
+
+        end_index = index + copyright_index + len(copyright) + \
+            len(copyright_year)
+        if copyright_year[-4:] != str(int(current_year) - 1):
+            # append current year
+            content = content[:end_index] + ', ' + current_year + \
+                content[end_index:]
+        elif copyright_year[-5:-4] == '-':
+            # update end year of range
+            content = content[:end_index - 4] + current_year + \
+                content[end_index:]
+        else:
+            # update year to be a range
+            content = content[:end_index] + '-' + current_year + \
+                content[end_index:]
+
+        # output beginning of file for debugging
+        # index = end_index
+        # for _ in range(3):
+        #     index = get_index_of_next_line(content, index)
+        # print('<<<')
+        # print(content[:index - 1])
+        # print('>>>')
+
+        with open(filename, 'w') as h:
+            h.write(content)
+
+
+def add_copyright(filename, copyright):
+    """
+    Add the copyright message to a file.
+
+    The copyright is placed below an optional shebang line as well as an
+    optional coding line.
+    It is preceded by an empty line if not at the beginning of the file and
+    always followed by an empty line.
+    """
+    with open(filename, 'r') as h:
+        content = h.read()
+
+    begin_index = scan_past_coding_and_shebang_lines(content)
+    end_index = scan_past_empty_lines(content, begin_index)
+
+    # inject copyright message
+    comment = get_comment(filename, copyright)
+    inserted_block = '%s\n\n' % comment
+    if begin_index > 0:
+        inserted_block = '\n' + inserted_block
+    content = content[:begin_index] + inserted_block + content[end_index:]
+
+    # output beginning of file for debugging
+    # index = end_index + len(inserted_block)
+    # for _ in range(3):
+    #     index = get_index_of_next_line(content, index)
+    # print('<<<')
+    # print(content[:index - 1])
+    # print('>>>')
+
+    with open(filename, 'w') as h:
+        h.write(content)
+
+
+def scan_past_coding_and_shebang_lines(content):
+    index = 0
+    while (
+        is_comment_line(content, index) and
+        (is_coding_line(content, index) or
+         is_shebang_line(content, index))
+    ):
+        index = get_index_of_next_line(content, index)
+    return index
+
+
+def scan_past_empty_lines(content, index):
+    while is_empty_line(content, index):
+        index = get_index_of_next_line(content, index)
+    return index
+
+
+def get_index_of_next_line(content, index):
+    index_n = content.find('\n', index)
+    index_r = content.find('\r', index)
```

Comment:
The problem is that the file can have any of these characters which does not need to correspond with the OS this script is running on.

I can also not use `splitlines()` or something similar here since I need the absolute index in the file.